### ISSUE TITLE
🌀[API] Adjust BorderBox Props

### ISSUE THREAD

**Issue by emplums** [primer/components#510]
- Remove default border radius
- Explore variants for common border types
- Make editing one of the borders easier

**Comment by emplums** [comment]
I noticed today while working on documentation that the `borderColor` prop is currently broken. Due to the structure of the component, the border color always gets overridden by the default.

**Comment by BinaryMuse** [comment]
I've been trying to work through the `BorderBox` issues recently and while there's been some conversation in Slack, I wanted to capture thoughts here.

Currently, `BorderBox` is broken in the following ways:

* Works:
    * `<BorderBox m={1} border="none">`
    * `<BorderBox m={1} borderRadius={3}>`
* Doesn't work:
    * `<BorderBox m={1} border="3px dashed black">` - color stays gray
    * `<BorderBox m={1} borderColor="green.5">` - border gets darker but doesn't turn green (?)
    * `<BorderBox m={1} borderWidth={5}>` - width doesn't change
    * `<BorderBox m={1} borderStyle="dashed">` - border style doesn't change

This is because we set two default props: `border="1px solid"` and `borderColor="gray.2"` and `styled-system` isn't smart enough to order props in the correct way.

## Splitting into individual default props

One thought is to split out the `border` props into individual props, so that we have the following default props:

```javascript
BorderBox.defaultProps = {
  borderWidth: '1px',
  borderStyle: 'solid',
  borderColor: 'gray.2',
  borderRadius: 2
}
```

This fixes the original issues, but breaks the ability to use any `border` or `border-<side>` shorthand properties:

* Works:
    * `<BorderBox m={1} borderColor="green.5">`
    * `<BorderBox m={1} borderWidth={5}>`
    * `<BorderBox m={1} borderStyle="dashed">`
    * `<BorderBox m={1} borderRadius={3}>`
* Doesn't work:
    * `<BorderBox m={1} border="none">`
    * `<BorderBox m={1} border="3px dashed black">`

It seems like folks use the `border` shorthand a lot when they don't want a box with borders on all four sides; take [this example](https://github.com/primer/doctocat/blob/e145052d0a3174ffb11f1685429024bc86a13975/theme/src/components/nav-items.js#L37-L43) from Doctocat (which is repeated in several places):

```javascript
 <BorderBox
  border={0}
  borderRadius={0}
  borderTop={1}
>
```

We also use this pattern in [our own docs](https://primer.style/components/SideNav#lightweight-variant) to get a bottom-border:

```javascript
<BorderBox border='none' borderBottom={1} mb={2} pb={1}>
  <Heading as="h5" fontSize={1} color="gray.7">Menu</Heading>
</BorderBox>
```

(Side note: there's a bug here because the border is still curved on the left and right due to the default border radius.)

With this set of default props, replicating that example (to get a bottom-border only) would require the above to be changed:

```javascript
<BorderBox borderWidth={0} borderBottomWidth={1} borderRadius={0} ...>
```

which I suppose is not all that bad, but it's probably not clear to an end-user why just setting `border="none"` like before doesn't work. Additionally, this would be a breaking change and all our internal components that use it as well as all apps which use `BorderBox` with any border customization at all would need to be changed.

## Other considerations

If we're looking at making a breaking change to `BorderBox` (and it seems like fixing this bug will necessitate a breaking change), this may be a good time to revisit this component.

### `BorderBox` as a content container

For rendering a container of text that is wrapped in a GitHub-style border (e.g. comment containers, the box around the Markdown editors, etc.), this component is doing its job well.

### `BorderBox` in layout

`Flex` is extremely useful for building page-scale layouts, but often one wants a border to separate parts of the UI and `Flex` doesn't support border props. The solution is to reach for the only building-block component Primer Components offers, `BorderBox`. It could be argued that one should simply create their own `styled-component` driven container with the right styles, but I feel like Primer Components has the ability to provide a component that will fit the bill in most circumstances.

### `BorderBox` as decoration

It's relatively common to want a border on one edge of a component to separate content. An example of this is the `Heading` example, above. In this case, the `BorderBox` defaults aren't that useful. You have to disable the default border, apply a border to the edge you want, and remember to turn off the border radius.

## Solution brainstorm

One possibility is to add a prop to `BorderBox` that disables the default props. Given the right theme values, one could write something like:

```javascript
<BorderBox reset borderBottom="borders.underline">
  <Heading>...</Heading>
</BorderBox>
```

A better solution might be to split out components so that `BorderBox` isn't used for so many things. For example,

* `<ContentBox><Text>This is styled like the current BorderBox</Text></ContentBox>`
    The default `BorderBox` works well for content, so this would be basically a name change
* `<UnderlineBox><Heading>A heading</Heading></UnderlineBox>`
* <BorderBox><Text>This is just a box with

**Comment by BinaryMuse** [comment]
I've been trying to work through the `BorderBox` issues recently and while there's been some conversation in Slack, I wanted to capture thoughts here.

Currently, `BorderBox` is broken in the following ways:

* Works:
    * `<BorderBox m={1} border="none">`
    * `<BorderBox m={1} borderRadius={3}>`
* Doesn't work:
    * `<BorderBox m={1} border="3px dashed black">` - color stays gray
    * `<BorderBox m={1} borderColor="green.5">` - border gets darker but doesn't turn green (?)
    * `<BorderBox m={1} borderWidth={5}>` - width doesn't change
    * `<BorderBox m={1} borderStyle="dashed">` - border style doesn't change

This is because we set two default props: `border="1px solid"` and `borderColor="gray.2"` and `styled-system` isn't smart enough to order user-supplied props in the correct way to work with the CSS cascade. See https://github.com/styled-system/styled-system/issues/1058 and the linked issues there for more information on how people are running into this bug.

## Splitting into individual default props

One thought is to split out the `border` props into individual props, so that we have the following default props:

```javascript
BorderBox.defaultProps = {
  borderWidth: '1px',
  borderStyle: 'solid',
  borderColor: 'gray.2',
  borderRadius: 2
}
```

This fixes the original issues, but breaks the ability to use any `border` or `border-<side>` shorthand properties:

* Works:
    * `<BorderBox m={1} borderColor="green.5">`
    * `<BorderBox m={1} borderWidth={5}>`
    * `<BorderBox m={1} borderStyle="dashed">`
    * `<BorderBox m={1} borderRadius={3}>`
* Doesn't work:
    * `<BorderBox m={1} border="none">`
    * `<BorderBox m={1} border="3px dashed black">`

It seems like folks use the `border` shorthand a lot when they don't want a box with borders on all four sides; take [this example](https://github.com/primer/doctocat/blob/e145052d0a3174ffb11f1685429024bc86a13975/theme/src/components/nav-items.js#L37-L43) from Doctocat (which is repeated in several places):

```javascript
 <BorderBox
  border={0}
  borderRadius={0}
  borderTop={1}
>
```

We also use this pattern in [our own docs](https://primer.style/components/SideNav#lightweight-variant) to get a bottom-border:

```javascript
<BorderBox border='none' borderBottom={1} mb={2} pb={1}>
  <Heading as="h5" fontSize={1} color="gray.7">Menu</Heading>
</BorderBox>
```

(Side note: there's a bug here because the border is still curved on the left and right due to the default border radius.)

With this set of default props, replicating that example (to get a bottom-border only) would require the above to be changed:

```javascript
<BorderBox borderWidth={0} borderBottomWidth={1} borderRadius={0} ...>
```

which I suppose is not all that bad, but it's probably not clear to an end-user why just setting `border="none"` like before doesn't work. Additionally, this would be a breaking change and all our internal components that use it as well as all apps which use `BorderBox` with any border customization at all would need to be changed.

## Other considerations

If we're looking at making a breaking change to `BorderBox` (and it seems like fixing this bug will necessitate a breaking change), this may be a good time to revisit this component.

### `BorderBox` as a content container

For rendering a container of text that is wrapped in a GitHub-style border (e.g. comment containers, the box around the Markdown editors, etc.), this component is doing its job well.

### `BorderBox` in layout

`Flex` is extremely useful for building page-scale layouts, but often one wants a border to separate parts of the UI and `Flex` doesn't support border props. The solution is to reach for the only building-block component Primer Components offers, `BorderBox`. It could be argued that one should simply create their own `styled-component` driven container with the right styles, but I feel like Primer Components has the ability to provide a component that will fit the bill in most circumstances.

### `BorderBox` as decoration

It's relatively common to want a border on one edge of a component to separate content. An example of this is the `Heading` example, above. In this case, the `BorderBox` defaults aren't that useful. You have to disable the default border, apply a border to the edge you want, and remember to turn off the border radius.

## Solution brainstorm

One possibility is to add a prop to `BorderBox` that disables the default props. Given the right theme values, one could write something like:

```javascript
<BorderBox reset borderBottom="borders.underline">
  <Heading>...</Heading>
</BorderBox>
```

A better solution might be to split out components so that `BorderBox` isn't used for so many things. For example,

* `<ContentBox><Text>This is styled like the current BorderBox</Text></ContentBox>`
    The default `BorderBox` works well for content, so this would be basically a name change
* `<UnderlineBox><Heading>A heading</Heading></UnderlineBox>`
    This would serve the common pattern of adding a bottom border with no radius
* `<BorderBox><Text>This is just a box with border props</Text></BorderBox>`
    This would simply expose the border system props with no default styles

---

This is about as far as I've gotten in my current investigation; definitely interested in people's thoughts. 💭

**Comment by dmarcey** [comment]
Not sure if this is in the scope of the discussion as it's not related to `<BorderBox />` in particular, more to borders, but I thought I would at least throw it out there.

Should there be a more "first-class" way to disable / modify borders of existing components? For example, if I want to create a "borderless" version of `<TextInput />`, it would be nice if there were a way to say `<TextInput border={0} />`. Instead, we end up doing something like:

```
export const BorderlessTextInput = styled("input")<FontSizeProps & FontWeightProps & LineHeightProps>`
    border: 0;

   ...

    ${typography}
`;
```

At this point we're no longer building onto of `<TextInput />` so if there is anything else from that component that we want in our `<BorderlessTextInput />` we're forced to duplicate these styles. Saying that this is something that is outside of the system and this is how we want to ask consumers to achieve the desired effect (no borders), is a reasonable stance, I just wanted to call it out as a potential pain point where I was curious if the design system could provide this sort of flexibility.

If `border` were to become a common prop, then that could possibly solve this problem, but I realize that is quite a heavy-handed approach.

Also, if this isn't relevant to the conversation, feel free to ignore for the purposes of this discussion. It seemed at least adjacent to the topic, so I thought I would mention it 😅

**Comment by colebemis** [comment]
Thanks for starting this discussion, @binarymuse! If I understand correctly, you've identified two problems with the `BorderBox` API: 

### Problem 1

 Mixing the shorthand `border` prop with longhand props like `borderColor` and `borderWidth` can lead to unexpected and unpredictable results.

### Problem 2
`BorderBox` has inconvenient default prop values that are often "reset" by consumers. Specifically, by default, `BorderBox` has a 1px border on all sides and a border radius. To create a box with a border on only one side—which is a common use case—you have to reset the border width of the other three sides and reset the border radius.  

---

Here are my thoughts:

### Addressing problem 1

I'm okay with removing the shorthand `border` prop and forcing consumers to use the longhand `borderWidth`, `borderStyle`, and `borderColor` props. As you mentioned, removing the `border` prop might create some initial confusion. However, I think removing the `border` prop is easier than fighting with the CSS cascade and will create a more predictable API. 

### Addressing problem 2

I think we should remove the default border width and border radius from `BorderBox` and instead create a variant for the "GitHub-style" box. That way creating a box with a border on only one side doesn't involve reset several irrelevant props and, if you do want a "GitHub-style" box, you can still make one with having to guess the correct border styles:

```diff
// Box with a bottom border
- <BorderBox border="none" borderBottom={1} bordeRadius={0}></BorderBox>
+ <BorderBox borderBottomWidth={1}></BorderBox>

// GitHub-style content box
- <BorderBox></BorderBox>
+ <BorderBox variant="content"></Box> 
//           ^^^^^^^^^^^^^^^^^                
// I'm not sure what we'd want to call this variant, but hopefully you get the idea.
```

**Comment by broccolini** [comment]
Here's my rushed feedback, apologies for not getting into finer detail, but wanted to plop here before our Code Review call.

**General**
Make the sx prop available to all components like or as part of common props. If a developer wants to override the core styles of a component then that’s what the sx prop is for. It’s a safe escape hatch that uses the theme object so we don’t have to worry about too many unexpected consequences. We still maintain encapsulation and themeability.

**Layout**
Give Box the flex props. Give the Flex component display: flex by default (and the other stuff we discussed with flex items, I can’t remember all the deats but seems like we came to good solutions a few weeks back and doesn’t look like it’s been implemented.

With the sx prop being available to every component, when someone wants to add a border that doesn’t suit the patterns provided by BorderBox, then they’ll be able to add easily.

**BorderBox**
BorderBox should work for the same examples as Primer CSS Box component. This is to create a rounded border box, that can optionally include a header, list of bordered rows, and footer. This component should work with lists or div so that it is semantically correct. Consider props for common variants like Danger or density. If a user doesn’t use those props, common props should work for spacing etc. If they want to make overrides to border props, they should use the sx prop.

Potentially this component could be used as a foundation for a more complex table-like component, or for lists with more complex headers like the issues list that has bulk actions and sorting. Only works if it closely follows those patterns, and only likely to need small exceptions that can be overridden by the SX prop.

This should reduce or eliminate cascade issues because SX prop can be passed for overrides, and the component shouldn’t be used if not for it’s core purpose.

In Theme UI for example, the SX prop is available everywhere, it’s a wrapper around the create element component that turns jsx into a classname.

**Comment by BinaryMuse** [comment]
> and the other stuff we discussed with flex items, I can’t remember all the deats but seems like we came to good solutions a few weeks back and doesn’t look like it’s been implemented

I can't recall the specifics of this conversation either. @emplums? If nothing else I can go look through old videos.

**Comment by BinaryMuse** [comment]
Fixed by #761, merged and ready for release in `v19.0.0`.
